art_tree.go

```diff
@@ -17,40 +20,90 @@ func NewArtTree() *ArtTree {
 	return &ArtTree{root: nil, size: 0}
 }
 
+type Result struct {
+	Key   string
+	Value interface{}
+}
+
+func (t *ArtTree) EachChanResult() chan Result {
+	return t.EachChanResultFrom(t.root)
+}
+
+func (t *ArtTree) EachChanResultFrom(start *ArtNode) chan Result {
+	outChan := make(chan Result)
+	go func() {
+		if start != nil {
+			for n := range t.EachChanFrom(start) {
+				if n.IsLeaf() {
+					outChan <- Result{string(n.key), n.value}
+				}
+			}
+		}
+		close(outChan)
+	}()
+	return outChan
+}
+
+// Finds the starting node for a prefix search and returns an array of all the objects under it
+func (t *ArtTree) PrefixSearch(key []byte) []interface{} {
+	ret := make([]interface{}, 0)
+	for r := range t.PrefixSearchChan(key) {
+		ret = append(ret, r.Value)
+	}
+	return ret
+}
+
+func (t *ArtTree) PrefixSearchChan(key []byte) chan Result {
+	return t.EachChanResultFrom(t.searchHelper(t.root, key, 0))
+}
+
 // Returns the node that contains the passed in key, or nil if not found.
 func (t *ArtTree) Search(key []byte) interface{} {
 	key = ensureNullTerminatedKey(key)
-	return t.searchHelper(t.root, key, 0)
+	foundNode := t.searchHelper(t.root, key, 0)
+	if foundNode != nil {
+		// i think with the null terminated key, the return is always a leaf, or nil?
+		// if foundNode.IsLeaf() {
```

Comment:
Correct, `searchHelper` will only ever return the resultant node or `nil` if it does not find a match.  LeafNodes are the only ones that can match a search term.

Would you mind removing this comment as a part of your PR?  Thanks!